splitpipe/doc/joinpipe.1.txt
JOINPIPE(1)
===========
bert hubert <ahu@ds9a.nl>
v0.2, 17 April 2005

NAME
----
joinpipe - joins volumes generated by splitpipe(1) to standard output

SYNOPSIS
--------
'joinpipe' 		[--help, -h] [--no-prompt, -n] [--verbose, -v]
		      	[--debug, -d] [--version] device1 file2 ...


DESCRIPTION
-----------
joinpipe(1) joins volumes generated by splitpipe(1) to standard output.
In doing so, it checks the proper order of inserted volumes, if these
volumes all belong to the same backup session. Integrity of data
is verified using a running checksum, currently MD5.

Volumes are read from the devices or filenames specified on the command
line. When passing multiple names, the first volume is read from the first
name, the second volume from the second name etc. When there are more
volumes than names, remaining volumes are read in from the last name
specified.

EXAMPLES
--------

To backup /home to one or more DVDs:

 # tar czf - /home | splitpipe -s dvd  -o 'growisofs -Z /dev/dvd=/dev/stdin'

To restore:

 # joinpipe /dev/dvd | tar xzf -

OPTIONS
-------
--help, -h::
	Print a helpful message

--version::
	Print version information

--no-prompt, -n::
	Do not prompt the user to change volumes. Mostly useful when writing
	to files, but might also come in handy with automated disk changers.

--verbose, -v::
	Be verbose, but not overly so. 

--debug, -d::
	Give large amounts of debugging output, possibly damaging
	performance

BUGS
----
Does not yet allow retrying failed input, for example, after inserting the
wrong volume.

AUTHOR
------
Written by Netherlabs Computer Consulting BV, bert hubert, <ahu@ds9a.nl>

RESOURCES
---------
Website: http://ds9a.nl/splitpipe

SEE ALSO
--------
splitpipe(1), splitpipe(5)

COPYING
-------
Copyright (C) 2005 Netherlabs Computer Consulting BV. Free use of this
software is granted under the terms of the GNU General Public License (GPL)
version 2.



splitpipe/CHANGELOG
Please see http://ds9a.nl/splitpipe/changelog.html


splitpipe/joinpipe.cc
/*
    joinpipe takes volumes output by splitpipe and joins them to one pipe
    Copyright (C) 2005  Netherlabs Computer Consulting BV

    This program is free software; you can redistribute it and/or modify
    it under the terms of the GNU General Public License version 2 as published by
    the Free Software Foundation;

    This program is distributed in the hope that it will be useful,
    but WITHOUT ANY WARRANTY; without even the implied warranty of
    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
    GNU General Public License for more details.

    You should have received a copy of the GNU General Public License
    along with this program; if not, write to the Free Software
    Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
*/

#include <unistd.h>
#include <getopt.h>
#include <fcntl.h>
#include <sys/select.h>
#include <time.h>
#include <sys/time.h>
#include <sys/wait.h>
#include <stdint.h>
#include <signal.h>
#include <iostream>
#include <netinet/in.h>
#include <stdexcept>
#include <vector>
#include "misc.hh"
#include "md5.hh"

using namespace std;

namespace {
  struct params
  {
    params()
    {
    verbose = debug = verify = noPrompt = false;
    }
    
    bool verbose;
    bool debug;
    bool verify;
    bool noPrompt;
    vector<string> inputDevice;
  }parameters;
}

static void usage()
{
  cerr<<"joinpipe joins multiple volumes (volumes) into one pipe.\n";
  cerr<<"\nsyntax: joinpipe [options] device1 [device2] | ...\n\n";
  cerr<<" --debug, -d\t\tGive debugging output\n";
  cerr<<" --help, -h\t\tGive this helpful message\n";
  cerr<<" --no-prompt, -n\tRun without user intervention\n";
  cerr<<" --verbose, -v\t\tGive verbose output\n";
  cerr<<" --version\t\tReport version\n\n";
  
  exit(1);
}

static void ParseCommandline(int argc, char** argv)
{
  int c;
  
  while (1) {
    int option_index = 0;
    static struct option long_options[] = {
      {"debug", 0, 0, 'd'},
      {"help", 0, 0, 'h'},
      {"verbose", 0, 0, 'v'},
      {"no-prompt", 0, 0, 'n'},
      {"version", 0, 0, 'e'},
      {0, 0, 0, 0}
    };
    
    c = getopt_long (argc, argv, "dhnv",
		     long_options, &option_index);
    if (c == -1)
      break;
    
    switch (c) {
    case 'd':
      parameters.debug=1;
      break;
    case 'e':
      cerr<<"joinpipe "VERSION" (C) 2005 Netherlabs Computer Consulting BV\nReport bugs to bert hubert <ahu@ds9a.nl>"<<endl;
      exit(EXIT_SUCCESS);
    case 'h':
      usage();
      break;

    case 'n':
      parameters.noPrompt=1;
      break;

    case 'v':
      parameters.verbose=true;
      break;
    case '?':
      usage();
      break;
    }
  }
  if (optind < argc) {
    while (optind < argc) {
      parameters.inputDevice.push_back(argv[optind++]);
    }
  }
  if(parameters.inputDevice.empty()) {
    cerr<<"Error: missing input device of filename(s)\n\n";
    usage();
  }

}

int JoinpipeMain(int argc, char** argv)
try
{
  ParseCommandline(argc, argv);

  char* buffer=new char[65536];
  struct stretchHeader stretch;
  string uuid;
  unsigned int numVolumes=0;

  if(parameters.inputDevice.empty())
    parameters.inputDevice.push_back("/dev/stdin");

  vector<string>::const_iterator inputIter=parameters.inputDevice.begin();

  int infd=open(inputIter->c_str(), O_RDONLY);
  if(infd < 0)
    unixDie("opening of "+*parameters.inputDevice.begin()+" for input");


  MD5Summer md5;
  string label;

  for(;;) {
    if(!readn(infd, &stretch, sizeof(struct stretchHeader), "read of stretch header"))
      throw runtime_error("EOF before end of session");

    stretch.size=ntohs(stretch.size);

    if(stretch.size && !readn(infd, buffer, stretch.size, "read of a stretch of input")) {
      cerr<<stretch.size<<endl;
      throw runtime_error("unexpected end of file during read of a stretch");
    }

    if(stretch.type==stretchHeader::SessionUUID) {    
      if(uuid.empty()) {
	uuid=string(buffer,buffer+stretch.size);
	cerr<<"UUID of this session is '"<<makeHexDump(uuid)<<"'"<<endl;
      } else {
	if(uuid != string(buffer,buffer+stretch.size)) {
	  cerr<<"This volume does not belong to the correct session, ";
	  cerr<<"uuid should be '"<<makeHexDump(uuid)<<"', is '"<<makeHexDump(string(buffer,buffer+stretch.size))<<"'"<<endl;
	  if(!label.empty())
	    cerr<<"This backup has label '"<<label<<"'"<<endl;
	  cerr<<"joinpipe: please insert proper volume"<<endl;
	  close(infd);
	  inputIter++;
	  if(!parameters.noPrompt)
	    waitForUser();
	  infd=open(inputIter->c_str(), O_RDONLY);
	  if(infd < 0)
	    unixDie("opening of "+ *inputIter+" for input");
	}
      }
    }
    else if(stretch.type==stretchHeader::VolumeNumber) {    
      uint16_t newVolumeNumber;
      memcpy(&newVolumeNumber, buffer, 2);
      newVolumeNumber = ntohs(newVolumeNumber);
      if(newVolumeNumber != numVolumes) {
	cerr<<"This is volume number "<<newVolumeNumber + 1<<", expected "<<numVolumes + 1<<", please retry"<<endl;
	close(infd);
	inputIter++;
	if(!parameters.noPrompt)
	  waitForUser();
	infd=open(inputIter->c_str(), O_RDONLY);
	if(infd < 0)
	  unixDie("opening of "+ *inputIter+" for input");
      }
      else {
	cerr<<"joinpipe: found volume "<<newVolumeNumber + 1<<", as expected"<<endl;
	numVolumes++;
      }
    }
    else if(stretch.type==stretchHeader::Data) {    
      if(!writen(1, buffer, stretch.size, "write of a stretch of input"))
	throw runtime_error("unexpected end of file on standard output");
      md5.feed(buffer, stretch.size);
    }
    else if(stretch.type==stretchHeader::VolumeEOF) {    
      close(infd);
      cerr<<"joinpipe: end of volume, change media to volume "<<numVolumes + 1<<" and press enter"<<endl;

      if(!parameters.noPrompt)
	waitForUser();

      if(inputIter + 1 != parameters.inputDevice.end())
	inputIter++;

      infd=open(inputIter->c_str(), O_RDONLY);
      if(infd < 0)
	unixDie("opening of "+ *inputIter+" for input");
    }
    else if(stretch.type==stretchHeader::SessionEOF) {    
      cerr<<"joinpipe: end of session"<<endl;
      break;
    }
    else if(stretch.type==stretchHeader::MD5Checksum) {    
      string md5sum=md5.get();
      if(md5sum==string(buffer, buffer + stretch.size))
	cerr<<"joinpipe: running checksum correct"<<endl;
      else {
	cerr<<"joinpipe: checksum incorrect, actual '"<<makeHexDump(md5sum)<<"', should be '"<<makeHexDump(string(buffer,buffer+stretch.size))<<"'" <<endl;
	exit(EXIT_FAILURE);
      }
    }
    else if(stretch.type==stretchHeader::SessionName) {    
      label=string(buffer, buffer + stretch.size);
      cerr<<"joinpipe: Session label '"<< label <<"'"<<endl;
    }
    else if(stretch.type==stretchHeader::VolumeDate && stretch.size==4) {    
      time_t then;
      memcpy(&then, buffer, 4);
      then=ntohl(then);

      struct tm tm;
      tm=*localtime(&then);
      char tbuf[80];
      strftime(tbuf,sizeof(tbuf),"%a %b %d %H:%M:%S %Z %Y", &tm);

      cerr<<"joinpipe: volume was started on "<<tbuf<<endl;
    }
    else {
      cerr<<"joinpipe: unknown stretch type "<<(int)stretch.type<<" of length "<< stretch.size <<endl;
      cerr<<makeHexDump(string(buffer, buffer+stretch.size))<<endl;
    }
  }
  return EXIT_SUCCESS;
}
catch(exception &e)
{
  cerr<<"Fatal: "<<e.what()<<endl;
  return EXIT_FAILURE;
}
